fmriprep/interfaces/fmap.py

```diff
@@ -497,6 +539,56 @@ def phdiff2fmap(in_file, delta_te, newpath=None):
     return out_file
 
 
+def phases2fmap(phase_files, magnitude_files, metadatas, newpath=None):
+    """calculates a phasediff from two phase images. Assumes monopolar
+    readout and that the second image in the list corresponds to the
+    longer echo time"""
+    import numpy as np
+    import nibabel as nb
+    from nipype.utils.filemanip import fname_presuffix
+    from copy import deepcopy
+
+    phasediff_file = fname_presuffix(phase_files[0], suffix='_phasediff',
+                                     newpath=newpath)
+    echo_times = [meta.get("EchoTime") for meta in metadatas]
+    if None in echo_times or echo_times[0] == echo_times[1]:
+        raise RuntimeError()
+    # Determine the order of subtraction
+    short_echo_index = echo_times.index(min(echo_times))
+    long_echo_index = echo_times.index(max(echo_times))
+
+    magnitude_image = magnitude_files[short_echo_index]
+    short_phase_image = phase_files[short_echo_index]
+    long_phase_image = phase_files[long_echo_index]
+
+    mag_img = nb.load(magnitude_image)
+    mag = mag_img.get_data().astype(np.float)
+    image0 = nb.load(short_phase_image)
+    phase0 = image0.get_data().astype(np.float)
+    image1 = nb.load(long_phase_image)
+    phase1 = image1.get_data().astype(np.float)
+
+    # Calculate fieldmaps
+    rad0 = np.pi * phase0 / 2048 - np.pi
+    rad1 = np.pi * phase1 / 2048 - np.pi
```

Comment:
Will phases always range from 0 to 4096?

Reply:
We're trying to recreate these scanning sequences to see if this is still true for the current Siemens product. This has been true for all the older versions of the sequence I've run into.

Reply:
Thanks for the diligence on this. Do you know if any other scanners produce phase1/phase2 fieldmaps? It would be good to see if this convention extends beyond Siemens.

Perhaps @neurolabusc might have the experience or know someone who does.

Reply:
I have a web page that is [relevant](https://www.mccauslandcenter.sc.edu/crnl/tools/fieldmap). For Siemens images I think the scaled range is -4096..4092 (scl_slope = 2.000000; scl_inter = -4096.000000). For Philips I think the scaled range is ~-500..496 for some systems and -pi..+pi for others. For GE I think the convention is to save real/imaginary images so you would use fslcomplex to get the phase map. As a good estimate, you can assume min..max should be scaled very close to -pi..+pi.

Reply:
Thanks for the rundown. If the slope and intercept are 2, -4096, and the original range is 0..4096, then this should be:

```Python
rad0 = np.pi * phase0 / 4096
rad1 = np.pi * phase1 / 4096
```

This is because `img.get_fdata()` already applies the scaling, so we should work from the scaled values, not the raw values.

I think we can ignore the GE case for now; that seems like it should probably not be considered "phase1"/"phase2" fieldmaps, and it's a job for BIDS to establish what the correct naming scheme is to indicate those files.

I think in the short term, we could have some a heuristic where we check for values above 500 and pi, to establish the correct denominator.

In the long term, I'm leery of putting too many heuristics for interpreting data into fMRIPrep that are not in the BIDS spec. We may want to consider a metadata field that unambiguously indicates the correct interpretation. (Though the heuristic would probably need to stay, to keep processing datasets that conform with BIDS 1.0.)

Reply:
I changed so the scaling is more general in 09316a9, but in my test data this causes a very small error because the maximum observed value is 4094 instead of 4096.

Reply:
Maybe we could leave that as a last-resort option, and see if we can figure out a simple rescaling (no shift) in most cases? For example, a range of -4096.0 to 4092.0 exactly fits Chris's description of Siemens, so a `np.pi / 4096` scaling seems like a better choice.

Reply:
The minimum value that came out of `.get_fdata()` was 0, though.

Reply:
I see. So I guess we need to watch for cases where the scale and intercept are missing.

Reply:
We scanned all the available GRE sequences on a prisma and uploaded them on openneuro in case you want to take a look. Here's the link: https://openneuro.org/datasets/ds001600. These were converted with a recent `dcm2niix`. 

I'm open to suggestions on how to handle these robustly. Also, sorry about the formatting mess. I ran a misconfigured yapf and am still putting things back.

Reply:
@mattcieslak can you send me the raw DICOM data and permission to distribute the DICOMs on the [dcm2niix](https://www.nitrc.org/plugins/mwiki/index.php/dcm2nii:MainPage) wiki? This would provide a nice reference set for all DICOM-to-BIDS conversion tools to test, and it would provide nice validation for any future changes or features.

Reply:
Despite being acquired on a E11C Prisma that supports high dynamic range (16-bit DICOM), all the samples you acquired appear to use 12-bit precision (0..4095). My guess is 12-bit is more than sufficient for the SNR provided, but I would make sure that any code generates an error if image intensities are outside this range. This will detect any sequences that provide more than 12-bit precision.

- sub-1_acq-v1_magnitude1	0..1821
- sub-1_acq-v1_magnitude2	0..1705
- sub-1_acq-v1_phase1	0..4095
- sub-1_acq-v1_phase2	0..4095
- sub-1_acq-v2_magnitude1	0..1835
- sub-1_acq-v2_magnitude2	0..1738
- sub-1_acq-v2_phase1	0..4095
- sub-1_acq-v2_phase2	0..4095
- sub-1_acq-v4_magnitude1	0..1598
- sub-1_acq-v4_magnitude2	0..1566
- sub-1_acq-v4_phasediff	0..4095
- sub-1_dir-AP_epi	0..2341
- sub-1_dir-PA_epi	0..1796

Reply:
@neurolabusc sent a link to your mailbox.sc.edu from my pennmedicine email